### PR TITLE
[FIX] website_sale_loyalty: recalculate reward after delivery line deletion

### PR DESCRIPTION
**Pull request by anko-odoo**
If user claimed ewallet in confirm order stage, it would redirect them back to review order stage with wrong calculation if shipping had cost.
opw-3825880

### CODE REVIEW

File: addons/sale_loyalty_delivery/models/sale_order.py [hunk 1]
```diff
@@ -63,3 +63,8 @@ def _get_claimable_rewards(self, forced_coupons=None):
                     filtered_res[coupon] = filtered_rewards
             res = filtered_res
         return res
+
+    def _remove_delivery_line(self):
+        """Recalculate reward value if delivery line is deleted"""
```
Comment (Demesmaeker) [inline addons/sale_loyalty_delivery/models/sale_order.py:68]:
```suggestion
        """Override of delivery to recalculate the reward after a delivery line is deleted"""
```

File: addons/website_sale_loyalty/controllers/main.py [hunk 2]
```diff
@@ -103,7 +103,8 @@ def claim_reward(self, reward_id, code=None, **post):
                 coupon = coupon_
                 if code == coupon.code and (
                     (program_sudo.trigger == 'with_code' and program_sudo.program_type != 'promo_code')
-                    or (program_sudo.trigger == 'auto' and program_sudo.applies_on == 'future')
+                    or (program_sudo.trigger == 'auto' and program_sudo.applies_on == 'future'
+                        and program_sudo.program_type not in ('ewallet', 'gift_card', 'loyalty'))
```
Comment (Demesmaeker) [inline addons/website_sale_loyalty/controllers/main.py:107]:
```suggestion
                    or (
                        program_sudo.trigger == 'auto' 
                        and program_sudo.applies_on == 'future'
                        and program_sudo.program_type not in ('ewallet', 'gift_card', 'loyalty')
                    )
```

Comment (Demesmaeker) [inline addons/website_sale_loyalty/controllers/main.py:107]: Is the `gift_card` also missing in master then?
Why is this necessary for this bugfix?

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello :duck: 

The fix lgtm, nonetheless, there are some improvements that could still be done:
- Could you rename the PR and the commit? Like this, it sounds like you claim the ewallet during the order confirmation, which isn't what is done here =]
- Could you maybe add a test?

Have a nice day!

File: addons/website_sale_loyalty/controllers/main.py (hunk 2 above)
Comment (anko-odoo) [reply addons/website_sale_loyalty/controllers/main.py:107]: What do you mean?

Comment (Demesmaeker) [reply addons/website_sale_loyalty/controllers/main.py:107]: 1. This change is already present in master, but without the `gift_card`
2. I was wondering how this is related to this bugfix, as it's not exactly the same problem, this is mostly curiosity ^^

Comment (anko-odoo) [reply addons/website_sale_loyalty/controllers/main.py:107]: 1. The change in master was actually for a different thing and by accident fix half of the problem.
2. Indeed I checked again and the gift card one is different as with this fix I would break gift card totally. 
Will make take care of that one in different fix both for 16.3 and master.

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello, :duck: 

It's cleaner, indeed, and far better than before (nice catch for the kwargs ^^), but it raises another problem: when testing, after claiming the eWallet, I tried to change the shipping method to a more expensive one... The ewallet amount isn't changed (normal) but then... the amount to be paid staid at 0 too... '^^

1. 
image
2. 
image

Once I validated the order though:
image

:thinking: 

(also, could you please rewrite the commit messages, it's missing some determinant ^^)

**Comment by anko-odoo** [comment]
@Demesmaeker I know it's not a perfect solution and I will dig deeper into that, but I checked the case when the cheaper shipping wallet is able to cover the whole order, but with more expensive it would require some payment, then it works correctly and it refreshes the left amount. Also, i checked if this noncontinuity correctly subtracts points from the wallet, and it does. But indeed I will look closer to if it can be easily fixed to recalculate the value on the last step

**Review by chevalierv — COMMENTED** [review]
Hi!

I think this bug was solved in https://github.com/odoo/odoo/pull/160160.

Could you check and rebase your PR if it's not the case ?

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello @anko-odoo 

There are still some problems. Now, when paying with an eWallet, going to the delivery method step, everything looks fine, we still have the total amount to be paid,  etc. But if I select another delivery method, the paid with eWallet amount magically jumped to match the total order amount, without user consent, and the amount won't go down when selecting again the cheaper delivery method. '^^
IMHO, we shouldn't change the amount of the eWallet. Once the user select a delivery method, the user should again pay with the eWallet if he wants to also pay the delivery with it. In master, maybe we should consider eWallet as a payment method instead of allowing to add it before.

**Comment by anko-odoo** [comment]
@Demesmaeker Personally, I would find it less intuitive it would require user to reapply the ewallet everytime they change shipping method, as I suppose most people would prefer to use it if they have money on ewallet. @paju-odoo I think your could be valuable in that matter.
Also, could you provide me with more info regarding the case of choosing lower shipping, as when I am testing it, it seems to be fine.

**Comment by Demesmaeker** [comment]
When speaking about paying something, I'd hate seeing the amount paid by a method changing without an action from my side to allow them to use my money. But if we choose that, then it needs to be adapted when you go back to a cheaper method too. I'd not pay if it's written they take more from my eWallet than what is due for instance.
My step to have the problem in local:
- publish the post in website
- set the price to 15 for the standard delivery
- create an eWallet for Mitchell admin
- add something to the cart
- checkout
- pay with eWallet
- next step to choose delivery method. There everything work fine, still have an amount to be paid, could reapply eWallet
- select the post (20$ instead of 15$)
- eWallet price suddenly got updated to cover everything, nothing else to be paid
- go back to the standard delivery, total amount stay at 0, delivery price changed, eWallet amount didn't
- 
cc: @dalaOdoo too then, as it's a bit a in-between

**Comment by paju-odoo** [comment]
Hi @anko-odoo @Demesmaeker 
IMO, the user shouldn't have to reapply the ewallet code if the delivery amount changes. If you have enough money in your ewallet to cover the whole expense, you probably want to use it. I'd be annoyed to need to reapply the code when changing the delivery method and would probably miss it. 
Regarding the bug that @Demesmaeker mentionned, I tried to reproduce it on the runbot but did not encounter the issue. Maybe this was fixed in the meantime (?) :slightly_smiling_face:

File: addons/website_sale_loyalty/controllers/delivery.py [hunk 3]
```diff
@@ -24,4 +24,9 @@ def _update_website_sale_delivery_return(self, order, **post):
                         amount_free_shipping, currency
                     ),
                 })
+            else:
+                result.update({
+                    'new_amount_order_discounted': Monetary.value_to_html(
+                        order.reward_amount, {'display_currency': currency}),
+                })
```
Comment (Demesmaeker) [inline addons/website_sale_loyalty/controllers/delivery.py:31]:
```suggestion
                result.update({'new_amount_order_discounted': Monetary.value_to_html(
                    order.reward_amount, {'display_currency': currency}
                )})
```

**Review by Demesmaeker — APPROVED** [review]
Hello :duck: 

For the bug: that's strange, I rebased just before testing though... Anyway, if it's the expected comportment, then fine by me. @chevalierv if you want to look at it too :slightly_smiling_face: 

Have a nice day

**Review by chevalierv — APPROVED** [review]
LGTM :+1:

**Comment by Demesmaeker** [comment]
Hello :duck: 

Ready to g then ^^

@robodoo delegate+

@AntoineVDV Can we get your security higher powers as this is safe and controlled HTML?

**Comment by AntoineVDV** [comment]
@robodoo override=ci/security

**Comment by anko-odoo** [comment]
@robodoo r+

**Comment by anko-odoo** [comment]
@robodoo r+ rebase-ff
